### PR TITLE
Daemon concurrent rpc operations

### LINKED ISSUE #643: Handle RPC calls in parallel
We currently handle all RPC calls sequentially, meaning you have to wait for every long-running operation to complete before another one can be handled.

We need to decouple the RPC thread from the daemon thread and execute any long-running operations in separate threads again so the daemon is always responsive.

Plan of record is to peruse Qt's event loop, with the RPC thread posting events on the daemon's loop, which would handle them as appropriate - internally or via whatever mode of concurrency makes sense for the operation in question.

### PR DESCRIPTION
**Pull request by townsend2010**
Initial framework for making the daemon much more concurrent in how it handles rpc calls and the associated operations.

Also with this, the check for detecting if ssh is up in the instance is now asynchronous.

Fixes #643

### CODE REVIEW

**Comment by gerboland** [comment]
Overall I am happy with this and its direction.

I was hoping to avoid locking in VirtualMachine but for now it is clear it translates most directly and so is less likely to cause breakage. So +1 so far!

**Comment by townsend2010** [comment]
Hey guys!

Thanks for the initial reviews.  I agree that locking is rough, but it will be refined over time, ie, scoped locking of some sort (either c++17 `scoped_lock` or `lock_guard` in it's own scope block), who owns the mutex and does locking, etc.

Since it was mentioned, I don't think `async_future_watchers` needs any sort of locking.  The only thing that modifies and uses it is the daemon thread itself.  I do agree that there are probably other areas that need some locking...I just need to find those:)

**Comment by townsend2010** [comment]
Ok, I think this is ready for review for the first iteration of concurrency.

File: src/daemon/daemon_rpc.cpp [hunk 1]
```diff
@@ -84,6 +84,18 @@ auto make_server(const std::string& server_address, mp::RpcConnectionType conn_t
 
     return server;
 }
+
+template <typename OperationSignal>
+grpc::Status emit_signal_and_wait_for_result(OperationSignal operation_signal)
```
Comment (ricab) [inline src/daemon/daemon_rpc.cpp]: Ninja stuff, this!

File: src/daemon/daemon_rpc.cpp [hunk 2]
```diff
@@ -84,6 +84,18 @@ auto make_server(const std::string& server_address, mp::RpcConnectionType conn_t
...
+{
+    std::promise<grpc::Status> status_promise;
+    auto status_future = status_promise.get_future();
+    emit operation_signal(&status_promise);
+
+    status_future.wait();
```
Comment (ricab) [inline src/daemon/daemon_rpc.cpp]: This is a bit redundant since `get` already waits.

File: src/daemon/daemon.h [hunk 3]
```diff
@@ -60,7 +64,7 @@ struct MetricsOptInData
 };
 
 struct DaemonConfig;
-class Daemon : public QObject, public multipass::Rpc::Service, public multipass::VMStatusMonitor
+class Daemon : public QObject, public multipass::VMStatusMonitor
```
Comment (ricab) [inline src/daemon/daemon.h]: youch, good catch

File: src/daemon/daemon.h [hunk 4]
```diff
@@ -136,14 +140,32 @@ public slots:
...
+                                                std::promise<grpc::Status>* status_promise);
+    AsyncOperationStatus async_wait_for_ssh_all(const std::vector<std::string>& vms,
+                                                std::promise<grpc::Status>* status_promise);
+    template <typename Reply>
+    AsyncOperationStatus async_wait_for_ssh_and_start_mounts(grpc::ServerWriter<Reply>* server,
+                                                             const std::vector<std::string>& vms,
+                                                             std::promise<grpc::Status>* status_promise);
+    void finish_async_operation(QFuture<AsyncOperationStatus> async_future);
+    QFutureWatcher<AsyncOperationStatus>* create_future_watcher();
+    std::vector<std::unique_ptr<QFutureWatcher<AsyncOperationStatus>>> async_future_watchers;
```
Comment (ricab) [inline src/daemon/daemon.h]: Hmm, I am not sure I like this in the beginning... I understand you wanted to group it with related methods, but the division so far was between methods and attributes, and this moves all other attributes in the object by one position. Conceptually I also think it makes more sense to have config first, then instances, etc.

Was clang-format happy with the newline here?

File: src/rpc/multipass.proto [hunk 5]
```diff
@@ -280,7 +280,7 @@ message StartRequest {
 
 message StartReply {
     string log_line = 1;
-    string start_message = 2;
+    string reply_message = 2;
```
[the remark this answers is not in the document]
Comment (townsend2010) [reply src/rpc/multipass.proto]: @gerboland,

No, I don't think it's absolutely necessary, but I did this to make it more generic for when setting the reply message in the following piece if code:
https://github.com/CanonicalLtd/multipass/blob/a47a4e969a36dcbe9d6796c650a0e821bae8c1f9/src/daemon/daemon.cpp#L2142-L2143
`Reply` here is templated, but if left the RPC the way it was, it would be something like `reply.set_start_message(blah)` which didn't seem clean to me.

[the remark this answers is not in the document]
Comment (gerboland) [reply src/rpc/multipass.proto]: Fair enough!

File: src/daemon/daemon_rpc.cpp (hunk 2 above)
Comment (townsend2010) [reply src/daemon/daemon_rpc.cpp]: I'm not sure what you mean here.  Do you mean `get_future()` is supposed to wait (it doesn't)? Or do you mean there is a better method to use than `wait()`?

File: src/daemon/daemon.h (hunk 4 above)
Comment (townsend2010) [reply src/daemon/daemon.h]: Sure, I can move the attribute to the bottom of the other attributes and probably should have done that in the first place.  I'm pretty sure I ran `clang-format` on all of my commits, but I will again just to be sure.

File: src/daemon/daemon_rpc.cpp (hunk 1 above)
Comment (townsend2010) [reply src/daemon/daemon_rpc.cpp]: :)

File: src/daemon/daemon.h (hunk 3 above)
Comment (townsend2010) [reply src/daemon/daemon.h]: Yeah, it was cruft left over from when we split out the `DaemonRpc` class and should have been removed then.  But better late than never:)

File: src/daemon/daemon_rpc.cpp (hunk 2 above)
Comment (ricab) [reply src/daemon/daemon_rpc.cpp]: I was referring to the following get (line 98), which itself waits:

> [...] effectively calls wait() in order to wait for the result

(in https://en.cppreference.com/w/cpp/thread/future/get)

Comment (townsend2010) [reply src/daemon/daemon_rpc.cpp]: Ah, ok, lol, I didn't see the `get()` call within the comments and forgot that was there.  I will fix:)

**Comment by ricab** [comment]
> Sure, will do!

I meant here or over IRC, just to ease reviewing. I did not mean in the code, although I am fine with that too if you prefer.

> [...] as you saw, we don't need to inherit from `multipass::Rpc::Service` [...]

OK, so it was just a leftover param from the Rpc signatures that was propagated to all daemon stuff even thought it was never used, correct? We should drop [-Wno-unused-parameter](https://github.com/CanonicalLtd/multipass/blob/master/CMakeLists.txt#L88) to avoid this sort of thing...

File: src/daemon/daemon_rpc.cpp (hunk 2 above)
Comment (ricab) [reply src/daemon/daemon_rpc.cpp]: Yeah, github's review interface mixes everything up

**Comment by townsend2010** [comment]
Ok, I'll explain those lines here.

The `AsyncOperationStatus` struct is used to return the grpc `Statu`s along with the correct future that `DaemonRpc` created so the `Daemon` object can work on the correct operation when `finish_async_operation` is called.

`async_wait_for_ssh_for` is the base asynchronous method for the `wait_until_ssh_up` and only works on oneinstance.

`async_wait_for_ssh_all` is the asynchronous method for working on multiple instances and calls `async_wait_for_ssh_for` for each instance in a loop.  The loop itself is synchronous but since it's done in a separate thread, it's asynchronous as far as the daemon is concerned.

`async_wait_for_ssh_and_start_mounts` is the asynchronous method for starting an instance and does the `wait_until_ssh_up` and then sets up any mounts including installing `sshfs` in the instance if it's missing.

I hope this clears it up a bit.

Regarding removing `-Wno-unused-parameter`, I think we still want that there because `DaemonRpc` still overrides the `Rpc` class and `ServerContext` is not used there as well.

File: src/daemon/daemon.h (hunk 4 above)
Comment (townsend2010) [reply src/daemon/daemon.h]: FYI, moved the attribute to the bottom as well as running `clang-format` on all of the changes thus far.

File: src/daemon/daemon_rpc.cpp (hunk 2 above)
Comment (townsend2010) [reply src/daemon/daemon_rpc.cpp]: Done

**Comment by ricab** [comment]
OK, thanks for the explanation. So basically the \*async\* methods correspond to the long operations that should be offloaded, right?

On unused parameter, replying in #697

**Comment by townsend2010** [comment]
> So basically the *async* methods correspond to the long operations that should be offloaded, right?

Yep, exactly.

**Comment by gerboland** [comment]
I found a misbehaviour:

Terminal 1: Run `multipass launch --name x xenial` which started downloading the image (slow).
Terminal 2: Run `multipass launch --name y xenial`. It shows me "Creating y" spinner, then fails with

```
launch failed: failed to download from 'http://cloud-images.ubuntu.com/releases/server/releases/xenial/release-20190320/ubuntu-16.04-server-cloudimg-amd64-disk1.img': Network timeout
```

because the download in Terminal 1 was still in progress.

Terminal 1 download then failed with "launch failed: Cannot open image file for computing hash"



So maybe we need to check if 2 VMs are trying to launch simultaneously using the same download, and delay things?

**Comment by townsend2010** [comment]
@gerboland,

I kind of think this related to an already existing bug, #664. And also, the downloading of images part has not been touched by this PR.  I plan on making the image downloads better by making them asynchronous and also only try to download once.  I'd really like to defer that work though for the next iteration of this.

This PR is more about making `list`, etc. work while waiting for ssh to be ready during `launch`, `start`, and `restart`.

**Comment by gerboland** [comment]
@townsend2010 yep fine, I'm not asking for everything to be perfect in this MP.

File: src/daemon/daemon.cpp [hunk 6]
```diff
@@ -681,7 +682,7 @@ try // clang-format on
             reply.set_metrics_pending(true);
             server->Write(reply);
 
-            return grpc::Status::OK;
+            return status_promise->set_value(grpc::Status::OK);
```
Comment (ricab) [inline src/daemon/daemon.cpp]: Rather than changing all the code to replace every status return by the promise setting bit, I would prefer to see that encapsulated in a single place. For instance, a template that received a callable, forwarded the params to it, and set the return value in the promise. The signals would then be connected with instantiations of this template for launch, start, stop, etc. Changes would then be greatly simplified, the history of the code would be clearer, easier to review, and dozens of repetitions eliminated.

File: src/platform/backends/libvirt/libvirt_virtual_machine.cpp [hunk 7]
```diff
@@ -332,6 +342,16 @@ int mp::LibVirtVirtualMachine::ssh_port()
     return 22;
 }
 
+void mp::LibVirtVirtualMachine::ensure_vm_is_running()
+{
+    std::lock_guard<decltype(state_mutex)> lock{state_mutex};
+    if (domain_state_for(domain.get()) != VirtualMachine::State::running)
```
[the remark this answers is not in the document]
Comment (townsend2010) [reply src/platform/backends/libvirt/libvirt_virtual_machine.cpp]: In the qemu backend, we are checking if the qemu `QProcess` itself is running.  Here, we are relying on passing everything through `libvirtd` and so all we can do is check VM state.

File: src/platform/backends/libvirt/libvirt_virtual_machine.cpp [hunk 8]
```diff
@@ -287,17 +288,26 @@ void mp::LibVirtVirtualMachine::stop()
...
+    std::unique_lock<decltype(state_mutex)> lock{state_mutex};
     if (state == State::running || state == State::delayed_shutdown)
     {
         virDomainShutdown(domain.get());
         state = State::off;
         update_state();
     }
+    else if (state == State::starting)
+    {
+        virDomainDestroy(domain.get());
```
[the remark this answers is not in the document]
Comment (townsend2010) [reply src/platform/backends/libvirt/libvirt_virtual_machine.cpp]: Because we can assume if a `stop` is issued while an instance is starting, the we can just "pull the plug" on the VM and not worry about a graceful shutdown, which is what `virDomainShutdown()` does- a graceful shutdown.  In fact, if `virDomainShutdown()` is issued while the instance is booting, it is ignored.

File: src/platform/backends/libvirt/libvirt_virtual_machine.cpp [hunk 9]
```diff
@@ -332,6 +342,16 @@ int mp::LibVirtVirtualMachine::ssh_port()
     return 22;
 }
 
+void mp::LibVirtVirtualMachine::ensure_vm_is_running()
+{
+    std::lock_guard<decltype(state_mutex)> lock{state_mutex};
+    if (domain_state_for(domain.get()) != VirtualMachine::State::running)
+    {
+        state_wait.notify_all();
```
[the remark this answers is not in the document]
Comment (townsend2010) [reply src/platform/backends/libvirt/libvirt_virtual_machine.cpp]: What I'm trying to prevent is the daemon continuing it's operation until it's safe to so.  The situation I'm trying to avoid is a `multipass delete -p` on a starting instance and the daemon destroying the corresponding *VirtualMachine object while it's starting since the starting check is in another thread.  This is a synchronization between the daemon thread and the async_wait_until_ssh_up() thread.

File: src/platform/backends/libvirt/libvirt_virtual_machine.cpp [hunk 10]
```diff
@@ -378,10 +398,10 @@ std::string mp::LibVirtVirtualMachine::ipv6()
 
 void mp::LibVirtVirtualMachine::wait_until_ssh_up(std::chrono::milliseconds timeout)
 {
-    mp::utils::wait_until_ssh_up(this, timeout);
+    mp::utils::wait_until_ssh_up(this, timeout, std::bind(&LibVirtVirtualMachine::ensure_vm_is_running, this));
```
[the remark this answers is not in the document]
Comment (townsend2010) [reply src/platform/backends/libvirt/libvirt_virtual_machine.cpp]: Right, it's cruft left over from the awful process_events stuff we had before.  I debated changing the name to be more in line with what we really are doing here, but I decided at the time that this PR is very large and non trivial and really didn't want to add something somewhat unrelated.  I think I'll add a TODO here and fix in a follow up PR.

File: src/platform/backends/libvirt/libvirt_virtual_machine.cpp [hunk 11]
```diff
@@ -378,10 +398,10 @@ std::string mp::LibVirtVirtualMachine::ipv6()
...
 }
 
 void mp::LibVirtVirtualMachine::update_state()
 {
-    monitor->persist_state_for(vm_name);
+    monitor->persist_state_for(vm_name, state);
```
[the remark this answers is not in the document]
Comment (townsend2010) [reply src/platform/backends/libvirt/libvirt_virtual_machine.cpp]: I decided to do this because previously, `persist_state_for()` would turn right around and ask the vm for it's state, which really seems unnecessary given the vm is the one requesting this.  And also, having the vm tell the monitor (daemon) the state avoided a pretty serious concurrency issue I was running in to.

Also, `update_state()` is only called when new state is set, so state will always be outdated.

File: src/platform/backends/qemu/qemu_virtual_machine.cpp [hunk 12]
```diff
@@ -230,7 +230,7 @@ mp::QemuVirtualMachine::QemuVirtualMachine(const ProcessFactory* process_factory
                      [this](int exitCode, QProcess::ExitStatus exitStatus) {
                          mpl::log(mpl::Level::info, vm_name,
                                   fmt::format("process finished with exit code {}", exitCode));
-                         if (update_shutdown_status)
+                         if (update_shutdown_status || state == State::starting)
```
[the remark this answers is not in the document]
Comment (townsend2010) [reply src/platform/backends/qemu/qemu_virtual_machine.cpp]: `update_shutdown_status` is not new to this PR, so I think we should defer that discussion for later...unless I misunderstand what you mean here.

File: src/platform/backends/qemu/qemu_virtual_machine.cpp [hunk 13]
```diff
@@ -308,6 +308,9 @@ void mp::QemuVirtualMachine::shutdown()
     }
     else
     {
+        if (state == State::starting)
+            update_shutdown_status = false;
```
[the remark this answers is not in the document]
Comment (townsend2010) [reply src/platform/backends/qemu/qemu_virtual_machine.cpp]: Yeah, I think it's probably safe to remove this...probably some cruft left over.

File: src/daemon/daemon.cpp (hunk 6 above)
Comment (townsend2010) [reply src/daemon/daemon.cpp]: Ok, sure, I will look into that.

File: src/platform/backends/libvirt/libvirt_virtual_machine.cpp [hunk 14]
```diff
@@ -332,6 +342,16 @@ int mp::LibVirtVirtualMachine::ssh_port()
     return 22;
 }
 
+void mp::LibVirtVirtualMachine::ensure_vm_is_running()
```
[the remark this answers is not in the document]
Comment (townsend2010) [reply src/platform/backends/libvirt/libvirt_virtual_machine.cpp]: Because in the qemu backend, we are checking that the qemu's `QProcess` is actually running and didn't die for some reason.  Here, we are relying on `libvirtd` to do stuff and have no control of the qemu process itself.  I suppose it wouldn't hurt to add `ensure_vm_is_running()` in `LibVirtVirtualMachine::ssh_hostname()`, so I'll do that.

File: src/platform/backends/libvirt/libvirt_virtual_machine.cpp [hunk 15]
```diff
@@ -287,17 +288,26 @@ void mp::LibVirtVirtualMachine::stop()
...
         virDomainShutdown(domain.get());
         state = State::off;
         update_state();
     }
+    else if (state == State::starting)
+    {
+        virDomainDestroy(domain.get());
+        state = State::off;
+        update_state();
+        state_wait.wait(lock);
```
[the remark this answers is not in the document]
Comment (townsend2010) [reply src/platform/backends/libvirt/libvirt_virtual_machine.cpp]: See my answer in https://github.com/CanonicalLtd/multipass/pull/682#discussion_r271900060.

File: tests/test_libvirt_backend.cpp [hunk 16]
```diff
@@ -193,7 +193,7 @@ TEST_F(LibVirtBackend, machine_persists_and_sets_state_on_start)
     NiceMock<mpt::MockVMStatusMonitor> mock_monitor;
     auto machine = backend.create_virtual_machine(default_description, mock_monitor);
 
-    EXPECT_CALL(mock_monitor, persist_state_for(_));
+    EXPECT_CALL(mock_monitor, persist_state_for(_, _));
```
[the remark this answers is not in the document]
Comment (townsend2010) [reply tests/test_libvirt_backend.cpp]: I suppose we could, but will it gain us anything?

File: src/platform/backends/qemu/qemu_virtual_machine.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (townsend2010) [reply src/platform/backends/qemu/qemu_virtual_machine.cpp]: Oh wait, this is necessary.  The reason is when doing the `vm_process->kill()` after this, it triggers the `QProcess::errorOccurred` signal to be emitted, not the logic in line 233 that is contained in the slot when the `QProcess::finished` signal is emitted.

File: src/platform/backends/qemu/qemu_virtual_machine.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (ricab) [reply src/platform/backends/qemu/qemu_virtual_machine.cpp]: I guess my point is that changes here complexify an already intricate flag: `update_shutdown_status` used to mean approximately whether we want to record the shutdown or suspended state (IIUC). Now it doesn't anymore, because it may be `false` but we still want to do that. 

After your comment below, I understand the change is about not recording the shutdown state when there is an error while starting. Wouldn't it be better then to change the `if` in the error case to `if (update_shutdown_status && state != State::starting)` and leave `update_shutdown_state` alone here and below?

Additionally, in the error handler this flag is checked for a different purpose that what its name says. Couldn't the `if` there be based solely on state?

File: src/platform/backends/qemu/qemu_virtual_machine.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (ricab) [reply src/platform/backends/qemu/qemu_virtual_machine.cpp]: OK, that makes it much clearer. Please see reply above

File: src/platform/backends/libvirt/libvirt_virtual_machine.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (townsend2010) [reply src/platform/backends/libvirt/libvirt_virtual_machine.cpp]: Oh, forgot to address the other 2 parts:)

So, for spurious wakes, yes, I need to add a predicate.  I forgot about that, so will do.  Thanks!

For deadlocks, I'm unsure of your scenario.  There are also locks in the `action` and `on_timeout` lambdas in `mp::utils::wait_until_ssh_up()` so state won't be set or read unless the lock is being held.  I'll think about this some more just to make sure I'm not missing something.

File: src/platform/backends/libvirt/libvirt_virtual_machine.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (townsend2010) [reply src/platform/backends/libvirt/libvirt_virtual_machine.cpp]: `domain_state_for()` actually queries `libvirtd` for state and `libvirtd` will report `running` even when the VM is booting.  So if it's not running, then it's off and we take that as it was shutdown during start up.

For the `starting` in `shutdown`, this is the locally held state and not what `libvirtd` reports.

File: src/platform/backends/libvirt/libvirt_virtual_machine.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (townsend2010) [reply src/platform/backends/libvirt/libvirt_virtual_machine.cpp]: It's akin to pulling the plug on a physical machine during boot.  You can then push the power button again and it will boot.

File: src/platform/backends/qemu/qemu_virtual_machine.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (townsend2010) [reply src/platform/backends/qemu/qemu_virtual_machine.cpp]: Ok, I'll think about this some more and take your point into consideration.

File: src/daemon/daemon.cpp (hunk 6 above)
Comment (townsend2010) [reply src/daemon/daemon.cpp]: After thinking about this, I'm not really sure what you're wanting to achieve here.  Could you please provide a small example?

**Comment by gerboland** [comment]
I've spent some time testing this, and have not managed to break it. It's definitely a step in the right direction

File: src/daemon/daemon.cpp [hunk 17]
```diff
@@ -1392,81 +1400,30 @@ try // clang-format on
...
-
     if (config->update_prompt->is_time_to_show())
     {
         StartReply start_reply;
         config->update_prompt->populate(start_reply.mutable_update_info());
         server->Write(start_reply);
     }
 
-    return grpc_status_for(errors);
+    auto future_watcher = create_future_watcher();
```
[the remark this answers is not in the document]
Comment (townsend2010) [reply src/daemon/daemon.cpp]: Yep, you're right.

File: src/daemon/daemon.cpp [hunk 18]
```diff
@@ -2016,14 +1940,19 @@ grpc::Status mp::Daemon::create_vm(grpc::ServerContext* context, const CreateReq
...
         vm->start();
-        vm->wait_until_ssh_up(std::chrono::minutes(5));
 
         reply.set_vm_instance_name(name);
         config->update_prompt->populate_if_time_to_show(reply.mutable_update_info());
         server->Write(reply);
-    }
 
-    return grpc::Status::OK;
+        auto future_watcher = create_future_watcher();
```
[the remark this answers is not in the document]
Comment (townsend2010) [reply src/daemon/daemon.cpp]: yep

File: src/daemon/daemon.cpp (hunk 6 above)
Comment (ricab) [reply src/daemon/daemon.cpp]: I may be missing something, but my thought was: *since the promise is always being set to what was previously returned*, instead of changing each return to a promise setting call, we could have an intermediate thin layer to do that. This layer (let's call it `promise_fulfiller`) would sit between the rpc and `daemon::start`, `stop`, `launch`, etc. Those operations would still return the status (as until now), which the `promise_fulfiller` would use to set the promise. 

The `promise_fulfiller`  layer would probably be a template, receiving a method to call (one of `start`, `stop`, etc.), along with the corresponding parameters. The slots would not be `start`, `stop`, etc. directly, but an instantiation of `promise_fulfiller` with one of those methods.

Something along these lines (please treat this as pseudocode sketching):

```cpp
template<typename Operation, typename Request, typename Reply>
void promise_fulfiller(Operation op, Request request, Reply reply, std::promise promise)
{
  promise.set(op(request, reply));
}
...
QObject::connect(&rpc, &mp::DaemonRpc::on_launch, std::bind(promise_fulfiller, &mp::Daemon::launch));
```

**Comment by ricab** [comment]
OK, I just wrote a reply to clarify the sort of thing I had in mind in one of the review comments. As agreed in the meeting today, I am handing the review to Gerry now.

File: src/daemon/daemon.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (townsend2010) [reply src/daemon/daemon.cpp]: Actually, no, we want to keep this here the way it is.  It is the last thing to be printed out in the client.

File: src/daemon/daemon.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (townsend2010) [reply src/daemon/daemon.cpp]: Actually, no, it won't mess the order.  Only when `on_success` from the lambda in the client is called will this be printed out, so the full operation must finish.

File: src/daemon/daemon.cpp (hunk 6 above)
Comment (townsend2010) [reply src/daemon/daemon.cpp]: I see, interesting...I'll see about changing that.  Thanks!

Comment (townsend2010) [reply src/daemon/daemon.cpp]: Hmm, well, now that I think about it a bit more, I don't think this will work.  The operation itself doesn't necessarily set the value of the promise, but can be done in `mp::Daemon::finish_async_operation()` when there is an actual asynchronous operation.

File: src/daemon/daemon.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (townsend2010) [reply src/daemon/daemon.cpp]: Ok, I confirmed that this works as intended where it is, so no need to change.

File: src/daemon/daemon.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (townsend2010) [reply src/daemon/daemon.cpp]: Unlike `launch`, this does need fixing.

File: src/daemon/daemon.cpp (hunk 6 above)
Comment (gerboland) [reply src/daemon/daemon.cpp]: I agree with @townsend2010 here, it makes sense to permit a promise to be passed elsewhere, as it may take an operation on a separate thread to complete before the promise can be fulfilled, and we want the main loop unblocked asap.

File: src/platform/backends/libvirt/libvirt_virtual_machine.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (gerboland) [reply src/platform/backends/libvirt/libvirt_virtual_machine.cpp]: "Pull plug" made me worry, but libvirt does try to get the VM to shutdown properly:
https://libvirt.org/html/libvirt-libvirt-domain.html#virDomainDestroy
So this is good enough for me.

File: src/platform/backends/libvirt/libvirt_virtual_machine.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (gerboland) [reply src/platform/backends/libvirt/libvirt_virtual_machine.cpp]: LGTM

File: src/platform/backends/qemu/qemu_virtual_machine.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (gerboland) [reply src/platform/backends/qemu/qemu_virtual_machine.cpp]: I think @ricab has a point here, our state machine is getting hard to follow. It does need rethinking, but definitely not here.

I follow the intent of this line change, so I'll ok it.

File: src/platform/backends/qemu/qemu_virtual_machine.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (gerboland) [reply src/platform/backends/qemu/qemu_virtual_machine.cpp]: Understood

**Comment by townsend2010** [comment]
Bah, bors wants to use my comments in the, uh, comments section for the merge message.  I'm going to fix that up and re-run bors.

bors r-

**Comment by townsend2010** [comment]
bors r=ricab,gerboland
